### ISSUE TITLE
stats.linregress output inconsistences when x and y contains same values that is less than 1.

### ISSUE THREAD

**Issue by kkawabat** [scipy/scipy#10896]
I calculated linregress on two different pairs of array, where the first pair of array contains 1000 .0125 and the second pairs contain 10 .0125.
a = stats.linregress([.0125]*1000, [.0125]*1000)
b = stats.linregress([.0125]*10, [.0125]*10)

I would expect that the output slope and intercept to be the same for both however 
print(a.intercept, a.slope)
>> 0.0 1.0
print(b.intercept, b.slope)
>> nan nan

I tried to reproduce the problem with other values and it seems that if the value is between -1 and 1 and the length of array is larger than 1000 (or if the length of array is 3 for some reason) it will produce a slope of 1 and intercept of 0

for example
a = stats.linregress([-.0125]*3, [-.0125]*3)
a = stats.linregress([.099]*1040, [.099]*1040)
a = stats.linregress([.0125]*10000, [.0125]*10000)
will all produce: 
print(a.intercept, a.slope)
>> 0.0 1.0

#### Scipy/Numpy/Python version information:
`import sys, scipy, numpy; print(scipy.__version__, numpy.__version__, sys.version_info)`
>1.3.1 1.17.2 sys.version_info(major=3, minor=7, micro=4, releaselevel='final', serial=0)

**Comment by kkawabat** [comment]
The problem seems to be a rounding error in [np.cov](https://github.com/scipy/scipy/blob/c33dd0afbe45339a1dcc70511859457c36325bcf/scipy/stats/_stats_mstats_common.py#L116)

**Comment by tylerjereddy** [comment]
I'll tentatively label as a "defect" until a stats expert weighs in

**Comment by josef-pkt** [comment]
Looks like a corner case that doesn't make much sense. Correct answer should be an exception or (nan, nan),

The data has only a single point. We cannot find a unique solution to the regression problem, i.e. we are fitting a line to a single point and there is a continuum of lines that would fit the point perfectly.
Then the solution is just determined by floating point noise.

I guess the check to identify this case would have to be that cov is almost zero. (which requires a rough precision threshold)

**Comment by saskakarsi** [comment]
Could always also check if y.count(y[0]) == n and x.count(x[0]) == n, correct? Or for np arrays len(y[y==y[0]]) == n and len(x[x==x[0]]) == n. Eliminates the need for a precision check, could also make best-case pretty fast if you instead manually iterate through the array and break on the first element that is not x[0] or y[0]. Would love to submit a PR if this seems sensible!

**Comment by josef-pkt** [comment]
for a similar check of constant values, statsmodels uses `y.ptp()` or similar which was the fastest years ago in a stackoverflow question (but ambiguous speed comparison with recent numpy)

AFAIR, using `ptp` for nan checking had the problem that it raised a Warning if there are `invalid` values.

**Comment by saskakarsi** [comment]
Yep. What I'm saying is I think characteristic data that's input to a linreg is probably going to be disqualified rather quickly, need to of course ask whether speed of error checking is also critical. Quick illustrative example, used a million items to exaggerate effects

`
>>> def naiveloop(a):
...     for i in a[1:]:
...             if i != a[0]: break
`

`
>>> u = timeit.Timer("naiveloop(a)", setup="import numpy as np; a = np.random.rand(1000000); from __main__ import naiveloop")
>>> u.timeit(100)
7.316099799936637e-05
>>> u = timeit.Timer("naiveloop(a)", setup="import numpy as np; a = np.array([0.125]*1000000); from __main__ import naiveloop")
>>> u.timeit(100)
19.83621264500107
`
`
>>> u = timeit.Timer("np.any(a != a[0])", setup="import numpy as np; a = np.array([.1]*1000000)")
>>> u.timeit(100)
0.033780382000259124
>>> u = timeit.Timer("len(a[a==a[0]])", setup="import numpy as np; a = np.random.rand(1000000)")
>>> u.timeit(100)
0.04888108800150803
`
`
>>> u = timeit.Timer("np.ptp(a);", setup="import numpy as np; a = np.array([0.125]*1000000);")
>>> u.timeit(100)
0.05965185999957612
>>> u = timeit.Timer("np.ptp(a);", setup="import numpy as np; a = np.random.rand(1000000);")
>>> u.timeit(100)
0.06746767499862472
`

**Comment by mdhaber** [comment]
@tupui looks like gh-14757 fixed this, too. 
For the examples above, I get `ValueError: Cannot calculate a linear regression if all x values are identical`.

But really, this is again the same sort of catastrophic cancellation problem as addressed by gh-15905. If the values were slightly different, we wouldn't get the error, but we'd get garbage.

Shall we close it, though, since issue as originally reported is resolved and in this case the catastrophic cancellation occurs in NumPy?

**Comment by mdhaber** [comment]
Well, that PR alone will not make behavior in `linregress` more robust, but we can look into applying the idea here, too.
